### ISSUE TITLE
Egress gateway - Document requirement for destination on EC2

### ISSUE THREAD

**Issue by mikeinton** [cilium/cilium#16278]
## Bug report

Hi, we are attempting to test the newly released Egress IP NAT Gateway feature in 1.10 using this guide - https://docs.cilium.io/en/latest/gettingstarted/egress-gateway/.

After multiple attempts to schedule client pods on the same node as the gateway and also on a non gateway node, we are never able to actually get it to egress out of our configured `egressSourceIP`.

`cilium bpf egress list` returns no entries found from the source node (containing the client pod) which also happens to be the gateway node as detailed in the setup guide.

```
root@ip-96-102-39-42:/home/cilium# cilium bpf egress list
No entries found.
Note that for Linux kernel versions between 4.11 and 4.15 inclusive, the native
LPM map type used for implementing this feature does not provide the ability to
walk / dump the entries, so on these kernel versions this tool will never
return any entries, even if entries exist in the map.
```

- Cilium version `1.10.0 952d9d3`
- Kernel version `5.4.0-1045-aws`
- Orchestration system version in use `Kubernetes 1.20.6 deployed via Kops v1.21.0-beta1`
- Link to relevant artifacts (policies, deployments scripts, ...) `don't have a public link`
- Generate and upload a system zip: 

### Cilium config from Kops spec

```
  networking:
    cilium:
      agentPrometheusPort: 9090
      bpfCTGlobalAnyMax: 262144
      bpfCTGlobalTCPMax: 524288
      clusterName: default
      containerRuntimeLabels: none
      cpuRequest: 25m
      enableNodePort: true
      enableRemoteNodeIdentity: true
      hubble:
        enabled: false
      ipam: kubernetes
      memoryRequest: 128Mi
      monitorAggregation: medium
      sidecarIstioProxyImage: cilium/istio_proxy
      toFqdnsDnsRejectResponseCode: refused
      tunnel: vxlan
      version: v1.10.0
```

### Cilium Egress NAT Policy

We want to key off a specific namespace here
```
apiVersion: cilium.io/v2alpha1
kind: CiliumEgressNATPolicy
metadata:
  name: egress-sample
spec:
  egress:
  - namespaceSelector:
      matchLabels:
        ns: egress-a-public
  destinationCIDRs:
  - 96.102.39.104/32
  egressSourceIP: "96.102.39.58"
```

### Gateway Node IP ADDR

Confirming that our intended gateway node has one of our `egressSourceIP` added
```
2: ens5: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc mq state UP group default qlen 1000
    link/ether 02:18:98:ef:c4:32 brd ff:ff:ff:ff:ff:ff
    inet 96.102.39.42/27 brd 96.102.39.63 scope global dynamic ens5
       valid_lft 3521sec preferred_lft 3521sec
    inet 96.102.39.58/27 scope global secondary ens5
       valid_lft forever preferred_lft forever
    inet 96.102.39.59/27 scope global secondary ens5
       valid_lft forever preferred_lft forever
    inet 96.102.39.60/27 scope global secondary ens5
       valid_lft forever preferred_lft forever
    inet6 2001:558:fd00:603:58c9:bcd7:c8cf:c407/128 scope global dynamic noprefixroute
       valid_lft 390sec preferred_lft 90sec
    inet6 fe80::18:98ff:feef:c432/64 scope link
       valid_lft forever preferred_lft forever
```

### Cilium Status Output

Attached - [cilium-status-verbose.txt](https://github.com/cilium/cilium/files/6525384/cilium-status-verbose.txt)

**Comment by pchaigno** [comment]
Thanks for the report! You told me on Slack you enabled the egress gateway via the Cilium ConfigMap. Was that before or after loading the egress NAT policy?

**Comment by mikeinton** [comment]
Thanks @pchaigno.  These were added to the configmap before we started applying our egress NAT policies which we've also deleted/removed multiple times during tests.

```
enable-egress-gateway: true
enable-bpf-masquerade: true
kube-proxy-replacement: strict
```

**Comment by pchaigno** [comment]
Did you restart the Cilium pods after adding those to the ConfigMap?

**Comment by mikeinton** [comment]
Yes, we also checked the cilium agent pod logs to confirm that those configs were set.

```
level=info msg="  --enable-bpf-masquerade='true'" subsys=daemon
level=info msg="  --enable-egress-gateway='true'" subsys=daemon
level=info msg="  --kube-proxy-replacement='strict'" subsys=daemon
level=info msg="  --kube-proxy-replacement-healthz-bind-address=''" subsys=daemon
```

**Comment by mikeinton** [comment]
Updating here that when restarting the cilium agent pods, we do see this logged.  UID also matches to what we see when we describe the policy.

```
level=info msg="Added CiliumEgressNATPolicy" ciliumEgressNATPolicyName=egress-sample k8sApiVersion=cilium.io/v2alpha1 k8sUID=633d4189-6743-4617-bf6a-36259c294a25 subsys=k8s-watcher
```

Yet `cilium bpf egress list` remains empty.

**Comment by aanm** [comment]
@mikeinton do you have any endpoint in `cilium endpoint list` that has the label `io.cilium.k8s.namespace.labels.ns=egress-a-public`?

**Comment by mikeinton** [comment]
Hi @aanm, no this is the closest result.

```
2197       Disabled           Disabled          35311      k8s:class=mediabot                                                                    100.96.3.205   ready
                                                           k8s:io.cilium.k8s.policy.cluster=default
                                                           k8s:io.cilium.k8s.policy.serviceaccount=default
                                                           k8s:io.kubernetes.pod.namespace=egress-a-public
                                                           k8s:org=empire
```

Do I need to change my policy to use a `podSelector` instead of `namespaceSelector`?  This is what I followed to setup the original CiliumEgressNATPolicy - https://docs.cilium.io/en/latest/gettingstarted/egress-gateway/#create-egress-nat-policy.

**Comment by mikeinton** [comment]
I switched our policy to `podSelector` and not `cilium bpf egress list` works.

Updated Policy
```
apiVersion: cilium.io/v2alpha1
kind: CiliumEgressNATPolicy
metadata:
  name: egress-sample
spec:
  egress:
  - podSelector:
      matchLabels:
        org: empire
        class: mediabot
        io.kubernetes.pod.namespace: egress-a-public
  destinationCIDRs:
  - 96.102.39.104/32
  egressSourceIP: "96.102.39.58"
```

```
root@ip-96-102-39-42:/home/cilium# cilium bpf egress list
SRC IP & DST CIDR               EGRESS INFO
100.96.4.64 96.102.39.104/32    96.102.39.58 96.102.39.58
100.96.2.198 96.102.39.104/32   96.102.39.58 96.102.39.58
100.96.3.228 96.102.39.104/32   96.102.39.58 96.102.39.58
```

Is this output correct?  None of my 3 source pods in that namespace can connect to the `DST CIDR` specified in this policy after it has been applied.  It seems like something still isn't right, is the output of `EGRESS INFO` valid?

In my initial post, I did note that the IP has been added to our target gateway node and it is reachable (tested via ping from another host) so I am not sure why the connection fails.  Once the NAT policy is deleted, the connection works again and nginx logs the pod's node source IP.

Snip of original `ip addr` listing the secondary IP as valid on our target gateway node
```
inet 96.102.39.58/27 scope global secondary ens5
       valid_lft forever preferred_lft forever
```

**Comment by pchaigno** [comment]
We're making progress :slightly_smiling_face: The output of `cilium bpf egress list` now looks okay.

Could you try to trace one of the failing connections on the source node (with `cilium monitor` or with Hubble)?

**Comment by aanm** [comment]
> Hi @aanm, no this is the closest result.
> 
> ```
> 2197       Disabled           Disabled          35311      k8s:class=mediabot                                                                    100.96.3.205   ready
>                                                            k8s:io.cilium.k8s.policy.cluster=default
>                                                            k8s:io.cilium.k8s.policy.serviceaccount=default
>                                                            k8s:io.kubernetes.pod.namespace=egress-a-public
>                                                            k8s:org=empire
> ```
> 
> Do I need to change my policy to use a `podSelector` instead of `namespaceSelector`? This is what I followed to setup the original CiliumEgressNATPolicy - https://docs.cilium.io/en/latest/gettingstarted/egress-gateway/#create-egress-nat-policy.

The issue might be the namespace `egress-a-public` does not have the label `ns=egress-a-public` set.

**Comment by mikeinton** [comment]
@pchaigno

We never get a corresponding ACK.
```
root@ip-96-102-39-42:/home/cilium# cilium monitor | grep 96.102.39.104
level=info msg="Initializing dissection cache..." subsys=monitor
-> overlay flow 0xb32df2e8 identity 35311->unknown state new ifindex cilium_vxlan orig-ip 0.0.0.0: 100.96.3.228:46452 -> 96.102.39.104:80 tcp SYN
-> overlay flow 0x1fa82f90 identity 35311->unknown state new ifindex cilium_vxlan orig-ip 0.0.0.0: 100.96.3.228:46452 -> 96.102.39.104:80 tcp SYN
```

I need to figure out why we can't reach our destination when sourcing from our designated `egressSourceIP` but it works when sourcing from the node's primary IP.
```
root@ip-96-102-39-42:~# tcptraceroute -s 96.102.39.58 96.102.39.104
Selected device ens5, address 96.102.39.58, port 56879 for outgoing packets
Tracing the path to 96.102.39.104 on TCP port 80 (http), 30 hops max
 1  * * *
 2  * * *
 3  * * *
 4  * * *
 5  * * *
```

Will keep troubleshooting our instances/security groups and post back with results.

@aanm - Thanks, I thought it would key off the namespace with out us having to label it but that makes sense.  I'll adjust our desired pod/namespace selector accordingly.

**Comment by pchaigno** [comment]
Are you trying to redirect packets through a specific cluster node or do you just want them to leave the original source node with the `egressSourceIP`? The egress IP gateway allows both.

Is 96.102.39.58 assigned to an interface of the source pod's node? Is 96.102.39.104 an IP inside or outside the cluster?

**Comment by mikeinton** [comment]
@pchaigno - Both actually, the guide has the test client mediabot pod assigned to the same node that runs the `"egress-ip-assign"` deployment.  I have a mediabot pod scheduled on that node as well as another one that isn't the intended gateway node.

Yes, 96.102.39.58 is assigned to the same interface as our designated gateway node here which has the primary IP of 96.102.39.42.  96.102.39.104 is completely outside of the cluster, it's a standalone EC2 instance running nginx.

**Comment by mikeinton** [comment]
It turns out that in addition to running the `ip address add` egress-ip-assign deployment to set our secondary IP's on our gateway node, if you're running on EC2 you also need to add those IP's to the instance itself.

We were able to test this directly by sourcing various secondary IP's directly from our gateway node to the standalone instance running nginx.

```
Node's primary IP
curl --interface 96.102.39.42 http://96.102.39.104

Node's secondary IP
curl --interface 96.102.39.58 http://96.102.39.104
```

The secondary IP never worked until we added them to the gateway instance in the AWS console.

Confirmed working when testing from the client mediabot test pod on the same node as the gateway and also from a pod that was scheduled on a non-gateway node.

AWS Docs for multiple IP's scenarios - https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/MultipleIP.html
```
To assign a secondary private IPv4 address to a network interface

1. Open the Amazon EC2 console at https://console.aws.amazon.com/ec2/.
2. In the navigation pane, choose Network Interfaces, and then select the network interface attached to the instance.
3. Choose Actions, Manage IP Addresses.
4. Under IPv4 Addresses, choose Assign new IP.
5. Enter a specific IPv4 address that's within the subnet range for the instance, or leave the field blank to let Amazon select an IP address for you.
6. (Optional) Choose Allow reassignment to allow the secondary private IP address to be reassigned if it is already assigned to another network interface.
7. Choose Yes, Update.
```

**Comment by Harsh63870** [comment]
hi, I'd like to work on this.
The fix is docs-only: document that on AWS EC2, secondary egress IPs must be assigned to the EC2 network interface in the AWS API (not only on the node via `ip address add`). I'll add a note under Preliminary Considerations in the egress gateway guide and cross-reference it from the `egressIP` warning.

**Comment by ysksuzuki** [comment]
The Preliminary Considerations section in the Egress Gateway documentation already states that the gateway interfaces and IP addresses must be provisioned and configured by the operator based on the networking environment, and it explicitly uses AWS/EKS ENIs and IP addresses as an example so that AWS can route traffic to and from the gateway nodes.
https://docs.cilium.io/en/stable/network/egress-gateway/egress-gateway/#preliminary-considerations

The exact way to make an egress IP usable/routable is environment-specific, and documenting vendor-specific mechanisms in the generic Egress Gateway guide does not scale well. I also do not see evidence that this particular EC2 secondary private IP case has been a recurring source of confusion.

Given that, I think the current vendor-neutral wording is the right level of abstraction, and I do not think we need to add AWS-specific guidance to the Egress Gateway docs for this issue. Closing this as already covered by the existing documentation.

Feel free to reopen if there is evidence that this is a recurring source of confusion, or if the existing documentation is missing a generic requirement that applies across environments.
